src/peft/tuners/loha/layer.py

```diff
@@ -0,0 +1,427 @@
+            module_dropout (`float`): The dropout probability for disabling adapter during training.
+            init_weights (`bool`): Whether to initialize weights
+        """
+
+        self.r[adapter_name] = r
+        self.alpha[adapter_name] = alpha
+        self.scaling[adapter_name] = alpha / r
+        self.rank_dropout[adapter_name] = rank_dropout
+        self.module_dropout[adapter_name] = module_dropout
+
+        # Determine shape of LoHa weights
+        if isinstance(self, nn.Linear):
+            shape = tuple(self.weight.shape)
+        elif isinstance(self, nn.Conv2d):
+            use_effective_conv2d = use_effective_conv2d and self.kernel_size != (1, 1)
+            if use_effective_conv2d:
+                shape = (self.out_channels, self.in_channels, *self.kernel_size)
+            else:
+                shape = (self.out_channels, self.in_channels * self.kernel_size[0] * self.kernel_size[1])
+        else:
+            raise NotImplementedError(f"LoHa is not implemented for {type(self).__name__} layer")
+
+        # Create weights with provided shape
+        self.create_loha_parameters(adapter_name, r, shape)
+
+        # Initialize weights
+        if init_weights:
+            self.reset_loha_parameters(adapter_name)
+
+        # Move new weights to device
+        weight = getattr(self, "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                self.to(weight.device, dtype=weight.dtype)
+            else:
+                self.to(weight.device)
+        self.set_adapter(self.active_adapters)
+
+    def get_delta_weight(self, adapter_name: str) -> torch.Tensor:
+        # https://github.com/KohakuBlueleaf/LyCORIS/blob/eb460098187f752a5d66406d3affade6f0a07ece/lycoris/modules/loha.py#L178
+        if adapter_name in self.hada_t1.keys():
+            weight = make_weight_cp(
+                self.hada_t1[adapter_name],
+                self.hada_w1_a[adapter_name],
+                self.hada_w1_b[adapter_name],
+                self.hada_t1[adapter_name],
+                self.hada_w2_a[adapter_name],
+                self.hada_w2_b[adapter_name],
+                scale=torch.tensor(self.scaling[adapter_name]),
+            )
+        else:
+            weight = make_weight(
+                self.hada_w1_a[adapter_name],
+                self.hada_w1_b[adapter_name],
+                self.hada_w2_a[adapter_name],
+                self.hada_w2_b[adapter_name],
+                scale=torch.tensor(self.scaling[adapter_name]),
+            )
+
+        weight = weight.reshape(self.weight.shape)
+
+        # Perform rank dropout during training - drop rows of addition weights
+        rank_dropout = self.rank_dropout[adapter_name]
+        if self.training and rank_dropout:
+            drop = (torch.rand(weight.size(0)) > rank_dropout).float()
```

Comment:
```suggestion
            drop = (torch.rand(weight.size(0)) < rank_dropout).float()
```

Right?

Also, can we always cast to `float()` or would that be an issue with mixed precision? Would it be safer to cast to the dtype of `weight`?

Reply:
I almost also got tricked, but isn't it working as expected (as nn.Dropout works)?

```python
import torch
import torch.nn.functional as F

x = torch.ones(1000)
print((F.dropout(x, p=0.1) > 0).sum())
# tensor(897)

x = torch.rand(1000)
print((x > 0.1).sum())
# tensor(896)
```

So, as far as I know, `p` stands for the probability that the element will be replaced with zero, according to [torch docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.dropout.html).

Reply:
Ah, yes, you are right. My confusion stems from the fact that the variable is called `drop` but actually it signifies the activations that are _not_ dropped :) Maybe a better name would help.

Reply:
In terms of `float()` - yep, it seems reasonable.